firestore/src/Firestore.gs

```diff
@@ -0,0 +1,272 @@
+ */
+Firestore.prototype.getData = function(project, collection, schema, numResults) {
+  const documents = this.fetchDocuments(project, collection, numResults);
+    
+  var data = [];
+  const instance = this;
+  documents.forEach(function(document) {
+    var values = [];
+    // Provide values in the order defined by the schema.
+    schema.forEach(function(field) {
+      if (instance.PREDEFINED_FIELDS.indexOf(field.name) >= 0) {
+        var value = document[field.name];
+        switch(field.name) {
+          case 'createTime':
+          case 'updateTime':
+            values.push(instance.parseTimestamp(value));
+            break;
+          case 'name':
+            values.push(instance.parseDocumentId(value));
+            break;
+          default:
+            values.push(value);
+        }
+      } else {
+        const valueWrapper = document.fields[field.name];
+        if (valueWrapper) {
+          switch(field.dataType) {
+            case 'STRING':
+              if (field.semantics && field.semantics.semanticGroup == 'DATE_AND_TIME') {
+                values.push(instance.parseTimestamp(valueWrapper.timestampValue)); 
+              } else {
+                values.push(valueWrapper.stringValue);
+              }
+              break;
+            case 'NUMBER':
+              if (valueWrapper.integerValue != null) {
+                values.push(valueWrapper.integerValue);
+              } else {
+                values.push(valueWrapper.doubleValue);
+              }
+              break;
+            case 'BOOLEAN':
+              values.push(valueWrapper.booleanValue);
+              break;
+            default:
+              throw 'Unsupported data type: ' + field.dataType;    
+          }
+        } else {
+          values.push(null); 
+        }
+      }
+    });
+    data.push({
+      values: values
+    });
+  });
+  
+  return data;
+}
+
+
+/**
+ * Fetches documents from Firestore or cache.
+ *
+ * @param {string} project The Google Cloud project ID containing Firestore instance.
+ * @param {string} collection The Firestore collection to use.
+ * @param {number} numResults Maximum documents to fetch.
+ * @return {Array} Array of Firestore documents.
+ */
+Firestore.prototype.fetchDocuments = function(project, collection, numResults) {
+  const urlComponents = [
+    'https://firestore.googleapis.com/v1beta1/projects/',
+    project,
+    '/databases/(default)/documents/',
+    collection,
+    '?pageSize=',
+    this.PAGE_SIZE];
+  const url = urlComponents.join('');
+  
+  const cached = this.cache.get(url);
+  if (cached) {
+    return JSON.parse(cached); 
+  }
+  
+  var documents = [];
+  var token = null;
+  while (documents.length < numResults) {
+    var response = this.fetchPage(url, token);
+    if (response.documents && response.documents.length > 0) {
+      Array.prototype.push.apply(documents, response.documents);
+    }
+    if (response.nextPageToken && token != response.nextPageToken) {
```

Comment:
Does the last page have a nextPageToken pointing to itself? I'm a bit confused by this logic.

Reply:
Fixed. The second part of condition is unnecessary.